### PR TITLE
Issue 72: Implement RFC 0005-integer_overflow_systematic_checks

### PR DESCRIPTION
**Pull request by ericeberry**

### CODE REVIEW

**Review by mhabrat — CHANGES_REQUESTED** [review]
Have we used a static analyzer like Coverity to confirm all the potential overflow cases have been addressed?

File: reference/src/taimpl/include/porting/overflow.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @section Description
+ * @file overflow.h
+ *
+ * This file contains the functions and structures that check the addition operation for overflow.
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/include/porting/overflow.h:22]: I looked through the code. There are a few subtract and multiple operations but they all deal with small memory ranges like moving a memory pointer by 16 bytes or small number operations where the calculation would never overflow.

File: reference/src/taimpl/src/internal/cenc.c [hunk 2]
```diff
@@ -133,7 +134,18 @@ size_t cenc_get_required_length(
 
     size_t required_length = 0;
     for (size_t i = 0; i < subsample_count; i++) {
-        required_length += subsample_lengths[i].bytes_of_clear_data + subsample_lengths[i].bytes_of_protected_data;
+        size_t subsample_len;
+        if (add_overflow(subsample_lengths[i].bytes_of_clear_data, subsample_lengths[i].bytes_of_protected_data,
+                    &subsample_len)) {
+            return 0;
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/internal/cenc.c]: Used a #define

File: reference/src/taimpl/src/internal/cenc.c [hunk 3]
```diff
@@ -133,7 +134,18 @@ size_t cenc_get_required_length(
...
+        }
+
+        size_t new_required_length;
+        if (add_overflow(required_length, subsample_len, &new_required_length)) {
+            return 0;
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/internal/cenc.c]: Used a #define

**Comment by ericeberry** [comment]
Is coverity available for Github?

**Comment by mhabrat** [comment]
> Is coverity available for Github?
This is a question for RDK-M. I doubt it as I don't see any mention of Coverity on the RDK-M wiki. It looks like there may be some support for Cppcheck. If we really wanted to, we could run a standalone Coverity scan of tasecureapi. However, it would be a one-off/manual process.

**Review by mhabrat — APPROVED** [review]

**Pull request merged**